File: src/main/java/com/pubnub/api/managers/ReconnectionManager.java
package com.pubnub.api.managers;

import com.pubnub.api.PubNub;
import com.pubnub.api.callbacks.PNCallback;
import com.pubnub.api.callbacks.ReconnectionCallback;
import com.pubnub.api.enums.PNReconnectionPolicy;
import com.pubnub.api.models.consumer.PNStatus;
import com.pubnub.api.models.consumer.PNTimeResult;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

import java.util.Calendar;
import java.util.Random;
import java.util.Timer;
import java.util.TimerTask;


@Slf4j
public class ReconnectionManager {

    private static final int BASE_LINEAR_INTERVAL_IN_MILLISECONDS = 3000;
    private static final int MIN_EXPONENTIAL_BACKOFF = 2;
    private static final int MAX_EXPONENTIAL_BACKOFF = 32;

    private static final int BOUND = 1000;
    private static final int MILLISECONDS = BOUND;
    private static final int MAXIMUM_RECONNECTION_RETRIES_DEFAULT = 10;

    private ReconnectionCallback callback;
    private PubNub pubnub;

    private int exponentialMultiplier = 1;
    private int failedCalls = 0;

    private PNReconnectionPolicy pnReconnectionPolicy;
    private int maxConnectionRetries;
    private final Random random = new Random();

    /**
     * Timer for heartbeat operations.
     */
    private Timer timer;

    public ReconnectionManager(PubNub pubnub) {
        this.pubnub = pubnub;
        this.pnReconnectionPolicy = pubnub.getConfiguration().getReconnectionPolicy();
        this.maxConnectionRetries = pubnub.getConfiguration().getMaximumReconnectionRetries();
    }

    public void setReconnectionListener(ReconnectionCallback reconnectionCallback) {
        this.callback = reconnectionCallback;
    }

    public void startPolling() {
        if (isReconnectionPolicyUndefined()) {
            return;
        }

        exponentialMultiplier = 1;
        failedCalls = 0;

        registerRetryTimer();
    }

    private void registerRetryTimer() {
        // make sure only one timer is running at a time.
        stopHeartbeatTimer();

        if (isReconnectionPolicyUndefined()) {
            return;
        }

        if (!maxConnectionIsSetToInfinite() && failedCalls >= maxConnectionRetries) {
            callback.onMaxReconnectionExhaustion();
            return;
        }

        timer = new Timer("Reconnection Manager timer", true);

        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                callTime();
            }
        }, getNextIntervalInMilliSeconds());
    }

    private boolean maxConnectionIsSetToInfinite() {
        return maxConnectionRetries == -1;
    }

    int getNextIntervalInMilliSeconds() {
        int timerInterval = 0;
        failedCalls++;

        if (pnReconnectionPolicy == PNReconnectionPolicy.EXPONENTIAL) {
            exponentialMultiplier++;
            timerInterval = (int) (Math.pow(2, exponentialMultiplier) - 1);
            if (timerInterval > MAX_EXPONENTIAL_BACKOFF) {
                timerInterval = MIN_EXPONENTIAL_BACKOFF;
                exponentialMultiplier = 1;
                log.debug("timerInterval > MAXEXPONENTIALBACKOFF at: " + Calendar.getInstance().getTime());
            } else if (timerInterval < 1) {
                timerInterval = MIN_EXPONENTIAL_BACKOFF;
            }
            timerInterval = (int) ((timerInterval * MILLISECONDS) + getRandomDelayInMilliSeconds());
            log.debug("timerInterval = " + timerInterval + "ms at: " + Calendar.getInstance().getTime());
        }

        if (pnReconnectionPolicy == PNReconnectionPolicy.LINEAR) {
            timerInterval = (int) (BASE_LINEAR_INTERVAL_IN_MILLISECONDS + getRandomDelayInMilliSeconds());
        }
        return timerInterval;
    }

    private int getRandomDelayInMilliSeconds() {
        return random.nextInt(BOUND);
    }

    private void stopHeartbeatTimer() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }

    private void callTime() {
        pubnub.time().async(new PNCallback<PNTimeResult>() {
            @Override
            public void onResponse(PNTimeResult result, @NotNull PNStatus status) {
                if (!status.isError()) {
                    stopHeartbeatTimer();
                    callback.onReconnection();
                } else {
                    log.debug("callTime() at: " + Calendar.getInstance().getTime().toString());
                    registerRetryTimer();
                }
            }
        });
    }

    private boolean isReconnectionPolicyUndefined() {
        if (pnReconnectionPolicy == null || pnReconnectionPolicy == PNReconnectionPolicy.NONE) {
            log.warn("reconnection policy is disabled, please handle reconnection manually.");
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/com/pubnub/api/managers/ReconnectionManager.kt
package com.pubnub.api.managers

import com.pubnub.api.PNConfiguration
import com.pubnub.api.PubNub
import com.pubnub.api.callbacks.ReconnectionCallback
import com.pubnub.api.endpoints.Time
import com.pubnub.api.retry.RetryConfiguration
import org.slf4j.LoggerFactory
import java.util.Calendar
import java.util.Random
import java.util.Timer
import java.util.TimerTask
import kotlin.math.pow

internal class ReconnectionManager(val pubnub: PubNub) {

    private val log = LoggerFactory.getLogger("ReconnectionManager")

    private companion object {
        private const val LINEAR_INTERVAL = 3
        private const val MIN_EXPONENTIAL_BACKOFF = 1
        private const val MAX_EXPONENTIAL_BACKOFF = 32

        private const val MILLISECONDS = 1000
        private const val MAX_RANDOM_DELAY_IN_MILLIS = 1000
        private const val NO_RETRIES = 0
    }

    internal lateinit var reconnectionCallback: ReconnectionCallback

    private var exponentialMultiplier = 1
    private var failedCalls = 0

    private lateinit var pnReconnectionPolicy: RetryConfiguration
    private var maxConnectionRetries = 6

    private var timer: Timer? = null
    private val random = Random()

    internal fun startPolling(pnConfiguration: PNConfiguration) {
        pnReconnectionPolicy = pnConfiguration.retryConfForOldSubscribeLoop

        if (isReconnectionPolicyUndefined())
            return

        maxConnectionRetries = getMaxConnectionRetries(pnReconnectionPolicy)
        exponentialMultiplier = 1
        failedCalls = 0

        registerRetryTimer()
    }

    private fun getMaxConnectionRetries(retryConfiguration: RetryConfiguration): Int {
        return when (retryConfiguration) {
            is RetryConfiguration.None -> {
                NO_RETRIES
            }
            is RetryConfiguration.Linear -> {
                retryConfiguration.maxRetryNumber
            }
            is RetryConfiguration.Exponential -> {
                retryConfiguration.maxRetryNumber
            }
        }
    }

    private fun registerRetryTimer() {
        // make sure only one timer is running at a time.
        stopRetryTimer()

        if (isReconnectionPolicyUndefined()) {
            return
        }

        if (failedCalls >= maxConnectionRetries) {
            reconnectionCallback.onMaxReconnectionExhaustion()
            return
        }
        timer = Timer("Reconnection Manager timer", true)
        timer?.schedule(
            object : TimerTask() {
                override fun run() {
                    callTime()
                }
            },
            getBestIntervalInMillis().toLong()
        )
    }

    private fun getBestIntervalInMillis(): Int {
        var timerInterval = 0
        when (pnReconnectionPolicy) {
            is RetryConfiguration.Exponential -> {
                timerInterval = (2.0.pow(exponentialMultiplier.toDouble()) - 1).toInt()
                if (timerInterval > MAX_EXPONENTIAL_BACKOFF) {
                    timerInterval = MIN_EXPONENTIAL_BACKOFF
                    exponentialMultiplier = 1
                    log.info("timerInterval > MAXEXPONENTIALBACKOFF at: " + Calendar.getInstance().time.toString())
                } else if (timerInterval < 1) {
                    timerInterval = MIN_EXPONENTIAL_BACKOFF
                }
                timerInterval = (timerInterval * MILLISECONDS) + getRandomDelayInMilliSeconds()
                log.info("timerInterval = " + timerInterval + "ms at: " + Calendar.getInstance().time.toString())
            }
            is RetryConfiguration.Linear -> {
                timerInterval = (LINEAR_INTERVAL * MILLISECONDS) + getRandomDelayInMilliSeconds()
            }
            is RetryConfiguration.None -> {
                timerInterval = (LINEAR_INTERVAL * MILLISECONDS) + getRandomDelayInMilliSeconds()
            }
        }
        return timerInterval
    }

    private fun getRandomDelayInMilliSeconds(): Int {
        return random.nextInt(MAX_RANDOM_DELAY_IN_MILLIS)
    }

    private fun stopRetryTimer() {
        timer?.cancel()
    }

    private fun callTime() {
        // Time endpoint by default is retryable, here we don't want it to retry
        Time(pubnub = pubnub, excludeFromRetry = true)
            .async { result ->
                if (!result.isFailure) {
                    stopRetryTimer()
                    reconnectionCallback.onReconnection()
                } else {
                    log.info("callTime at ${System.currentTimeMillis()}")
                    exponentialMultiplier++
                    failedCalls++
                    registerRetryTimer()
                }
            }
    }

    private fun isReconnectionPolicyUndefined(): Boolean {
        if (pnReconnectionPolicy == RetryConfiguration.None) {
            log.info("reconnection policy is disabled, please handle reconnection manually.")
            return true
        }
        return false
    }
}
